library(readr)
library(magrittr)
library(dplyr)
library(ggplot2)
library(stringr)
library(reshape2)
library(tidyr)

#This script processes the raw F-plasmid pMLST data generated by ABRicate and summarised by `abricate --summary`
#Steps required:
#1: Convert empty cells, which actually contain dashes '-', to zeroes.
#2: Convert hits of 90% or greater to ones, thus giving binary presence or absence of each allele
#3: Remove FIC alleles
#4: Clean up the headers so they just include the letter and the number i.e. FII_2 = F2, FIA_2 = A2
#5: TRICKY PART: Melt or possibly some other function to give the allele combos for each strain in long format
#6: Break ties between multiple FII alleles
#6: Paste the alleles together to give the F-type

#Not in function for use later
'%notin%' <- Negate('%in%')

#Read in data
pMLST_raw <- 
  read_delim("ST58/ST58.F.pMLST.raw.csv",delim = ",", na = ".") 
#rename the #FILE column
pMLST_raw <- pMLST_raw %>% select(`#FILE`, SEQUENCE, GENE, `%COVERAGE`, `%IDENTITY`)
pMLST_raw <- rename(pMLST_raw, FILE = `#FILE`, COVERAGE = `%COVERAGE`, IDENTITY = `%IDENTITY`)
pMLST_raw$FILE <- gsub(".fasta", "", pMLST_raw$FILE)
#Filter to hits where coverage is 100 and ID >=90
pMLST_100_99 <- pMLST_raw %>% filter(COVERAGE >=100, IDENTITY >=90)
pMLST_100_99_cast <- dcast(pMLST_100_99, FILE + SEQUENCE ~ GENE, fun.aggregate = max, value.var = "IDENTITY")
pMLST_100_99_cast[pMLST_100_99_cast == -Inf] <- 0

p1 <- pMLST_100_99_cast

colnames(p1) <- gsub(x = colnames(p1), "FIB_", "B")
colnames(p1) <- gsub(x = colnames(p1), "FIA_", "A")
colnames(p1) <- gsub(x = colnames(p1), "FII_", "F")
colnames(p1) <- gsub(x = colnames(p1), "FIC_", "C")

#Split into separate alleles
FA <- p1 %>% select(FILE, contains("A"))
FB <- p1 %>% select(FILE, contains("B"))
FF <- p1 %>% select(FILE, SEQUENCE, contains("F"))

#MELT SUB-DFs
#A
FA_melt <- melt(FA, id.vars = c("FILE"), variable.name = "A-type")
FA_melt_filt <- FA_melt[FA_melt$value > 90,]
#Check that samples don't have more than one allele
length(unique(FA_melt_filt$FILE))

#B
FB_melt <- melt(FB, id.vars = "FILE", variable.name = "B-type")
FB_melt_filt <- FB_melt[FB_melt$value > 90,]
#Check that samples don't have more than one allele
length(unique(FB_melt_filt$FILE))

#F
FF_melt <- melt(FF, id.vars = c("FILE", "SEQUENCE"), variable.name = "F-type")
FF_melt_filt <- FF_melt[FF_melt$value > 90,]
#Check that samples don't have more than one allele
length(unique(FF_melt_filt$FILE))

#At thise point the FF data has multiple hits for the same strain but we want a single column of F-types
#Return filenames that appear twice
F_dup <- FF_melt_filt[duplicated(FF_melt_filt$FILE),]

#Lines to check these DFs had no duplicates - they should be length 0
#A_dup <- FA_melt_filt[duplicated(FA_melt_filt$FILE),]
#B_dup <- FB_melt_filt[duplicated(FB_melt_filt$FILE),]

#Filter the DF down to all the entries for names appearing twice
F_duplicates <- FF_melt_filt %>% filter(FILE %in% F_dup$FILE)
F_non_duplicates <- FF_melt_filt %>% filter(FILE %notin% F_dup$FILE)

#A number of samples have a 100 hit and a lower ID hit so drop everything that has less than 100
F_duplicates_ID_filter <- F_duplicates %>% filter(value == 100) 

#Get the remaining samples with two 100% hits
F_dup_100ID <- F_duplicates_ID_filter[duplicated(F_duplicates_ID_filter$FILE),]

#Filter the 100-100 hits out with the F_dup_filt_names variable and arrange them by name and contig. This yields 16 samples with two hits each
F_dup_100ID <- F_duplicates_ID_filter %>% filter(FILE %in% F_dup_100ID$FILE) %>% arrange(FILE, SEQUENCE)

#Get the samples that now have one allele instead of two
F_dup_not_100ID <- F_duplicates_ID_filter[!duplicated(F_duplicates_ID_filter$FILE),]
F_dup_not_100ID <- F_dup_not_100ID %>% filter(FILE %notin% F_dup_100ID$FILE) %>% select(FILE, `F-type`)


#Create a column designating each hit as A or B, for casting
split_vect <- rep(c("A", "B"), 16)
F_dup_100ID$F_split <- split_vect

#Cast the data so there are now columns for hits A and B, with the value being the allele
F_dup_solve <- dcast(data = F_dup_100ID, FILE ~ F_split, value.var = "F-type", drop = FALSE)
#Create a new column with the two hits pasted together
F_dup_solve <- F_dup_solve %>% mutate(`F-type` = paste(A, B, sep = "_")) %>% select(FILE, `F-type`)

F_non_duplicates2 <- F_non_duplicates %>% select(FILE, `F-type`)

FF_processed_dups <- rbind(F_non_duplicates2, F_dup_solve, F_dup_not_100ID)

#Stick it all together
FFA <- left_join(FF_processed_dups, FA_melt_filt %>% select(-value), by = "FILE")
FFAB <- left_join(FFA, FB_melt_filt %>% select(-value), by = "FILE") %>% select(FILE, `F-type`, `A-type`, `B-type`) 
FFAB <- apply(FFAB, 2, as.character)
FFAB <- as.data.frame(FFAB, stringsAsFactors = FALSE)
FFAB <- FFAB %>% replace_na(list(FILE = "-", `F-type` = "-", `A-type` = "-", `B-type` = "-"))

#Wooooohoooo 
F_type <- FFAB %>% mutate(full_type = paste(`F-type`, `A-type`, `B-type`, sep = ":"))


#Split into STs with 5 or more, and less than 5 representatives
major_F <- F_type %>% group_by(full_type) %>% filter(n() >= 10)
minor_F <- F_type %>% group_by(full_type) %>% filter(n() < 10)

#Designate STs with less than 5 as "Other"
minor_F$full_type <- "Other"

#Stick them back together
all_F_types <- rbind(major_F, minor_F)

all_F_types <- all_F_types %>% select(FILE, full_type) %>% rename(ID = FILE)

#Export tsv
write_tsv(all_F_types, "ST58/F_types.tsv")

#Draw itol table
create_itol_files("ST58/F_types.tsv")


